goopicsi.go

```diff
@@ -225,6 +226,68 @@ func NVMeControllerDisconnect(id int64) error {
 	return nil
 }
 
+// ExposeRemoteNVMe creates a new NVMe Subsystem and NVMe controller
+func ExposeRemoteNVMe(subsystemID string, subsystemNQN string, maxNamespaces int64, controllerID string) error {
+	if conn == nil {
+		err := dialConnection()
+		if err != nil {
+			return err
+		}
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	client := pb.NewFrontendNvmeServiceClient(conn)
+	data1, err := client.GetNVMeSubsystem(ctx, &pb.GetNVMeSubsystemRequest{SubsystemId: &pbc.ObjectKey{Value: subsystemID}})
+	if err != nil {
+		log.Printf("No existing NVMe Subsystem found with subsystemID: %v", subsystemID)
+	}
+
+	if data1 == nil {
+		response1, err := client.CreateNVMeSubsystem(ctx, &pb.CreateNVMeSubsystemRequest{
+			Subsystem: &pb.NVMeSubsystem{
+				Spec: &pb.NVMeSubsystemSpec{
+					Id:            &pbc.ObjectKey{Value: subsystemID},
+					Nqn:           subsystemNQN,
+					MaxNamespaces: maxNamespaces,
+				},
+			},
+		})
+		if err != nil {
+			log.Println(err)
+			return err
+		}
+		log.Printf("NVMe Subsytem created: %v", response1)
+	} else {
+		log.Printf("NVMe Subsystem is already present with the subsytemID: %v", subsystemID)
+	}
+
+	data2, err := client.GetNVMeController(ctx, &pb.GetNVMeControllerRequest{ControllerId: &pbc.ObjectKey{Value: controllerID}})
+	if err != nil {
+		log.Printf("No existing NVMe Controller found with controllerID %v:", controllerID)
+	}
+	if data2 == nil {
```

Comment:
why such a big `if` here ? just return immediately in case of controller found and then rest of the code is not requiring `if`

Reply:
This if condition is when the controller is not found corresponding to the given controllerID and thus we create a new NVMe controller. In case the controller is found, we are just logging that in line 287

Reply:
what I say is just reverse the condition and return first...
```
if (condition) {
  // a lot of code here
}
```
do
```
if (!condition) {
    log and return
}
// a lot of code here instead
```